### PR TITLE
Electrical refactor - part 2

### PR DESCRIPTION
**Pull request by nRiccobo**
The electrical-refactor has been an on-going process to merge new research capabilities, most importantly the high-voltage direct current (HVDC) export system. #112 merged most of the features into dev. This PR will merge the remaining features as well as some additional improvements including library cleanup, some updated core default values, and updates to the electrical design module. Along the way some doc files, pre-commit packages, and workflow files were updated too. Lastly, a detailed example was added to demonstrate how to select HVDC or HVAC for a generic project. 

These additions were added with the intention that the original ORBIT workflow would be unaffected. The additional HVDC capabilities were added to a new module and relies on the user to specify if they want to use it or not. An example. "Example - Using HVDC or HVAC.ipynb" was added to inform users. 

This class combines the elements of ExportSystemDesign() and the OffshoreSubstationDesign(). Its purpose is to represent the export system more accurately by linking the type of cable (AC versus DC) and substation’s components (i.e. transformers versus converters). Most export and substation component costs were updated to include a per-unit cost rather than a per-MW cost rate and they can be added to the yaml file as arguments too. Otherwise, those per-unit costs use default and were determined with the help of a subcontractor . 
•	This module’s components’ cost scales with number of cables and substations rather than plant capacity. 
•	The offshore substation cost is calculated based on the cable type and number of cables, rather than scaling function based on plant capacity. 
•	The mass of an HVDC and HVAC substation are assumed to be the same. Without any new information the substructure mass and cost functions did not change.
•	An experimental onshore cost function was also added to account for the duplicated interconnection components. Costs will vary depending on the cable type.

A new example, Example – Using HVDC or HVAC.ipynb, was added to showcase the capabilities of the ElectricalDesign() class. It demonstrates how to create projects using HVAC or HVDC cables and how to use ParametricManager to compare the two design decisions.

### CODE REVIEW

File: ORBIT/phases/design/electrical_export.py [hunk 1]
```diff
@@ -327,63 +378,87 @@ def substation_cost(self):
...
+        switchgear_cost = self._design.get("switchgear_cost", 4e6)
+
+        if "HVDC" in self.cable.cable_type:
             num_switchgear = 0
         else:
             num_switchgear = self.num_cables
-        self.switchgear_cost = num_switchgear * self._design.get(
-            "switchgear_cost", 134000
-        )
+
```
Comment (RHammond2) [inline ORBIT/phases/design/electrical_export.py:443]: This is an area where I'd traditionally make this a one-liner like `num_switchgear = 0 if "HVDC" in self.cable.cable_type else self.num_cables`.

File: ORBIT/phases/design/electrical_export.py [hunk 2]
```diff
@@ -327,63 +378,87 @@ def substation_cost(self):
...
+        if "HVDC" in self.cable.cable_type:
+            num_dc_breakers = self.num_cables
         else:
-            num_dc_breaker = self.num_cables
-        self.dc_breaker_cost = num_dc_breaker * self._design.get(
-            "dc_breaker_cost", 40000000
-        )  # 4e6
+            num_dc_breakers = 0
+
+        self.dc_breaker_cost = num_dc_breakers * dc_breaker_cost
```
Comment (RHammond2) [inline ORBIT/phases/design/electrical_export.py:461]: Same comment about one-liners.

File: ORBIT/phases/design/electrical_export.py [hunk 3]
```diff
@@ -492,32 +567,87 @@ def calc_topside_mass_and_cost(self):
...
         ----------
-        topside_fab_cost_rate : int | float
         topside_design_cost: int | float
         """
 
         _design = self.config.get("substation_design", {})
-        topside_fab_cost_rate = _design.get("topside_fab_cost_rate", 14500)
-        topside_design_cost = _design.get("topside_design_cost", 4.5e6)
+        # topside_fab_cost_rate = _design.get("topside_fab_cost_rate", 14500)
+        # topside_design_cost = _design.get("topside_design_cost", 4.5e6)
```
Comment (RHammond2) [inline ORBIT/phases/design/electrical_export.py:575]: Is this being preserved, or was it just missed in the cleanup?

File: ORBIT/phases/design/electrical_export.py [hunk 4]
```diff
@@ -492,32 +567,87 @@ def calc_topside_mass_and_cost(self):
...
         )
-        self.topside_cost = (
-            self.topside_mass * topside_fab_cost_rate + topside_design_cost
-        )
+        if self.cable.cable_type == "HVDC-monopole":
+            self.topside_cost = _design.get("topside_design_cost", 294e6)
+        elif self.cable.cable_type == "HVDC-bipole":
+            self.topside_cost = _design.get("topside_design_cost", 476e6)
+        else:
+            self.topside_cost = _design.get("topside_design_cost", 107.3e6)
```
Comment (RHammond2) [inline ORBIT/phases/design/electrical_export.py:586]: Great use case for a match-case statement if we were on Python 3.10+

File: ORBIT/phases/design/electrical_export.py [hunk 5]
```diff
@@ -492,32 +567,87 @@ def calc_topside_mass_and_cost(self):
...
+        else:
+            self.onshore_converter_cost = 0
+            self.onshore_switchgear_cost = self.num_cables * _switchgear_cost
+            self.onshore_construction = self.num_substations * _design.get(
+                "onshore_construction_rate", 5e6
+            )
+            self.onshore_compensation_cost = (
+                self.num_cables * _compensation_rate
+                + self.onshore_shunt_reactor_cost
+            )
```
Comment (RHammond2) [inline ORBIT/phases/design/electrical_export.py:641]: Something to think about for a 2.0 timeframe is how we could change default value setups like this to be a bit more streamlined and avoid this big `if/elif/else` blocks altogether.

File: ORBIT/phases/design/electrical_export.py [hunk 6]
```diff
@@ -492,32 +567,87 @@ def calc_topside_mass_and_cost(self):
...
+            + self.onshore_switchgear_cost
+            + self.onshore_construction
+            + self.onshore_compensation_cost
             + self.mpt_cost
-            + self.switchgear_cost
         )
 
         self._outputs["export_system"][
-            "onshore_construction_cost"
+            "onshore_substation_costs"
```
Comment (RHammond2) [inline ORBIT/phases/design/electrical_export.py:652]: Is this naming going to be backwards compatible with existing ORBIT users, or will it break things?

File: tests/phases/design/test_electrical_design.py [hunk 7]
```diff
@@ -116,31 +113,80 @@ def test_dc_oss_kwargs():
         assert cost != base_cost
 
 
+def test_new_old_hvac_substation():
```
Comment (RHammond2) [inline tests/phases/design/test_electrical_design.py:116]: Thanks for updating the tests to account for old and new functionality!

**Review by RHammond2 — COMMENTED** [review]
Thanks for putting this all together! It's great that the docstrings and documentation got expanded a bit to include more details. I left a handful of comments, but they're pretty minor.

File: ORBIT/phases/design/electrical_export.py [hunk 8]
```diff
@@ -34,20 +56,26 @@ class ElectricalDesign(CableSystem):
...
         },
         "substation_design": {
-            "mpt_cost_rate": "USD/MW (optional)",
+            "substation_capacity": "MW (optional)",
+            "num_substations": "int (optional)",
+            "mpt_cost": "USD/cable (optional)",
             "topside_fab_cost_rate": "USD/t (optional)",
             "topside_design_cost": "USD (optional)",
-            "shunt_cost_rate": "USD/MW (optional)",
+            "shunt_cost_rate": "USD/cable (optional)",
```
[the remark this answers is not in the document]
Comment (nRiccobo) [reply ORBIT/phases/design/electrical_export.py:64]: That is fair. I'll double check those before the final merge.

File: ORBIT/phases/design/electrical_export.py [hunk 9]
```diff
@@ -327,63 +378,87 @@ def substation_cost(self):
             + self.switchgear_cost
             + self.converter_cost
             + self.dc_breaker_cost
-            + self.ancillary_system_cost
+            + self.ancillary_system_costs
```
[the remark this answers is not in the document]
Comment (nRiccobo) [reply ORBIT/phases/design/electrical_export.py:381]: I think it was a typo and I only caught it when comparing the two modules.

File: ORBIT/phases/design/electrical_export.py (hunk 6 above)
Comment (nRiccobo) [reply ORBIT/phases/design/electrical_export.py:652]: Yes. I believe it is now but let me revisit it to confirm. By changing it to substation_costs, we avoid an issue in `export.py`

https://github.com/WISDEM/ORBIT/blob/8eb149eeb2926a98a813c73eb9aec40385009724/ORBIT/phases/install/cable_install/export.py#L174-L180

There is some existing code that counts the transmissions construction that would get overwritten if we specified these costs as "onshore_construction_cost". I'll make sure that is what we want.

File: ORBIT/phases/design/electrical_export.py [hunk 10]
```diff
@@ -16,8 +16,30 @@ class ElectricalDesign(CableSystem):
...
+        only require a single type of cable.
+    total_length : float
+        Total length of cable required to trasmit power.
+    total_mass : float
+        Total mass of cable required to transmit power.
+    sections_cables : np.ndarray, shape: (`num_cables, )
+        An array of `cable`.
+    sections_lengths : np.ndarray, shape: (`num_cables, )
+        An array of `length`.
+
```
[the remark this answers is not in the document]
Comment (nRiccobo) [reply ORBIT/phases/design/electrical_export.py:39]: Thanks. I'll probably give it another pass by the time its ready for main. That way it updates all the api documentation.

File: ORBIT/phases/design/electrical_export.py [hunk 11]
```diff
@@ -327,63 +378,87 @@ def substation_cost(self):
...
 
         touchdown = self.config["site"]["distance_to_landfall"]
+        shunt_cost_rate = self._design.get("shunt_cost_rate", 1e4)
 
-        if (
-            self.cable.cable_type == "HVDC-monopole"
-            or self.cable.cable_type == "HVDC-bipole"
-        ):
-            compensation = 0
+        _num_shunts = self.num_cables
```
[the remark this answers is not in the document]
Comment (nRiccobo) [reply ORBIT/phases/design/electrical_export.py:418]: I think I added it to be more consistent with the descriptions in the documentation. "This model assumes one shunt reactor for each export cable". I can condense this part into the line it is actually used to eliminate these assignments.

File: ORBIT/phases/design/electrical_export.py (hunk 5 above)
Comment (nRiccobo) [reply ORBIT/phases/design/electrical_export.py:641]: Yeah. This section has some overlap with `export.py`. There is some room to clean up this.

File: ORBIT/phases/design/electrical_export.py (hunk 3 above)
Comment (nRiccobo) [reply ORBIT/phases/design/electrical_export.py:575]: I guess it was preserved, but it probably doesn't need to be preserved this we use this in `oss_design.py`

File: ORBIT/phases/design/electrical_export.py (hunk 1 above)
Comment (nRiccobo) [reply ORBIT/phases/design/electrical_export.py:443]: I like it.

File: ORBIT/phases/design/electrical_export.py (hunk 6 above)
Comment (nRiccobo) [reply ORBIT/phases/design/electrical_export.py:652]: Goes with the comment above. There is some clean up to be had around the onshore cost before v2.0. However, I did check that nothing is double counted and made sure that it is backwards compatible.

File: ORBIT/phases/design/electrical_export.py (hunk 3 above)
Comment (RHammond2) [reply ORBIT/phases/design/electrical_export.py:575]: I'm just checking, though by the time it gets to the main branch we might want to clean it up. This is definitely your call, I just wanted to understand the rationale.

**Pull request merged**